kobo/apps/openrosa/apps/logger/tests/test_form_submission.py

```diff
@@ -244,12 +244,45 @@ def test_duplicate_submission_with_different_content(self):
         self.assertEqual(self.response.status_code, 201)
         self.assertEqual(Instance.objects.count(), pre_count + 1)
         inst = Instance.objects.order_by('pk').last()
-        self._make_submission(duplicate_xml_submission_file_path)
-        self.assertEqual(self.response.status_code, 201)
-        self.assertEqual(Instance.objects.count(), pre_count + 2)
+        with self.assertRaises(Exception):
+            self._make_submission(duplicate_xml_submission_file_path)
+        self.assertEqual(self.response.status_code, 409)
+        self.assertEqual(Instance.objects.count(), pre_count + 1)
         # this is exactly the same instance
         another_inst = Instance.objects.order_by('pk').last()
-        self.assertNotEqual(inst.xml, another_inst.xml)
+        self.assertEqual(inst.xml, another_inst.xml)
+
+    def test_duplicate_submission_with_same_content_but_with_attachment(self):
+        """
+        Test that submitting the same XML content twice,
+        first without and then with an attachment,
+        results in a single instance with the attachment added.
+        """
+        xml_submission_file_path = os.path.join(
+            os.path.dirname(__file__),
+            "../fixtures/tutorial/instances/tutorial_with_attachment",
+            "tutorial_2012-06-27_11-27-53_w_attachment.xml"
+        )
+        media_file_path = os.path.join(
+            os.path.dirname(__file__),
+            "../fixtures/tutorial/instances/tutorial_with_attachment",
+            "1335783522563.jpg"
+        )
+        initial_instance_count = Instance.objects.count()
+
+        # Test submission with XML file
+        self._make_submission(xml_submission_file_path)
+        initial_instance = Instance.objects.last()
+        self.assertEqual(self.response.status_code, 201)
+        self.assertEqual(Instance.objects.count(), initial_instance_count + 1)
+
+        # Test duplicate submission with attachment
+        with open(media_file_path, 'rb') as media_file:
+            self._make_submission(xml_submission_file_path, media_file=media_file)
+        self.assertEqual(self.response.status_code, 201)
+        self.assertEqual(Instance.objects.count(), initial_instance_count + 1)
+        self.assertEqual(Attachment.objects.filter(instance=initial_instance).count(), 1)
```

Comment:
I have added a new test for this. I had to remove `existing_attachment.file_hash == hash_attachment_contents(f.read())` from the logic so that it does not create a new attachment if the file name is the same but the content is different. I'm thinking about the edit scenario (with deprecatedID). Feeling a little sleepy as it’s quite late on my end, but I will definitely test this thoroughly :-)